subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java

```diff
@@ -216,12 +218,18 @@ private void visitArtifacts(IdeDependencyVisitor visitor) {
                     Set<ResolvedArtifactResult> javaDoc = auxiliaryArtifacts.get(componentIdentifier, JavadocArtifact.class);
                     javaDoc = javaDoc != null ? javaDoc : Collections.<ResolvedArtifactResult>emptySet();
                     visitor.visitModuleDependency(artifact, sources, javaDoc, isTestConfiguration(configurations.get(artifactIdentifier)));
+                } else if (isGradleApiDependency(artifact)) {
+                    visitor.visitGradleApiDependency(artifact, sourcesJarFactory.resolveGradleApiSources(artifact.getFile()), isTestConfiguration(configurations.get(artifactIdentifier)));
                 } else {
                     visitor.visitFileDependency(artifact, isTestConfiguration(configurations.get(artifactIdentifier)));
                 }
             }
         }
 
+        private boolean isGradleApiDependency(ResolvedArtifactResult artifact) {
+            return artifact.getFile().getName().startsWith("gradle-api") || artifact.getFile().getName().startsWith("gradle-test-kit");
```

Comment:
I don't like having to depend on string jar name prefixes but could not find a better way to spot the artifacts that need sources.